### PR TITLE
Collect local variables

### PR DESCRIPTION
**Pull request by if1live**
## Overview
I add new option, capture_locals. (https://github.com/getsentry/raven-python/blob/133fcdb74869bb270f75642889b17b0752bf0b6e/raven/conf/defaults.py#L54-L55)

If config.capture_locals is true, raven capture local variables and send to sentry.
See #7. [binding_of_caller](https://github.com/banister/binding_of_caller) gem. 

If config.capture_locals is false, use previuse behavior.

Default value of capture_locals is false.
Because I can't believe that code run in all environments. (too many ruby implements.)
We can disable capture_locals if not working.

## Related Issue
* #7
* #261

### CODE REVIEW

File: lib/raven/base.rb [hunk 1]
```diff
@@ -14,12 +14,49 @@
...
+          callers.each_with_index do |caller_obj, idx|
+            e.backtrace[idx].instance_variable_set(:@stack_info, caller_obj)
+          end
+        end
+        original_raise.call(e)
+      end
+    end
+  end
+
+  @@monkey_patch_raise_occur = false
```
Comment (nateberkopec) [inline lib/raven/base.rb]: Are there threading issues with using a class variable here?

File: lib/raven/base.rb [hunk 2]
```diff
@@ -14,12 +14,49 @@
...
+require 'binding_of_caller'
 
 major, minor, patch = RUBY_VERSION.split('.').map(&:to_i)
 if (major == 1 && minor < 9) || (major == 1 && minor == 9 && patch < 2)
   require 'raven/backports/uri'
 end
 
+module Kernel
+  def self.monkey_patch_raise
+    if @@monkey_patch_raise_occur == true
```
Comment (nateberkopec) [inline lib/raven/base.rb]: Hmmm... I see why you're doing this, but I wonder if there's a better way.

File: lib/raven/base.rb (hunk 1 above)
Comment (if1live) [inline lib/raven/base.rb]: 1. add @@monkey_patch_raise_mutex.synchronize { ... } 
2. check flag twice (outside of critical section, inside of critical section)

File: sentry-raven.gemspec [hunk 3]
```diff
@@ -17,6 +17,7 @@ Gem::Specification.new do |gem|
 
   gem.add_dependency "faraday", ">= 0.7.6"
   gem.add_dependency "certifi"
+  gem.add_dependency "binding_of_caller" if RUBY_VERSION >= '1.9.2' && !defined?(JRUBY_VERSION)
```
Comment (eugeneius) [inline sentry-raven.gemspec]: The gemspec is evaluated when the gem is built, not when it's installed - `RUBY_VERSION` here will return the version of Ruby that is being used to build the gem. RubyGems doesn't support conditional dependencies.

I don't know if there's a way to maintain compatibility with Ruby 1.8.7 and also optionally depend on binding_of_caller.

Comment (nateberkopec) [inline sentry-raven.gemspec]: I think we should remove this line, and instead, people should be instructed to add "binding_of_caller" to their Gemfile if they want to use this feature. Similar to how `dalli` uses `kgio`.

Comment (eugeneius) [inline sentry-raven.gemspec]: That sounds good!

Dalli tries to require `kgio` at startup, but rescues the `LoadError` if it fails: https://github.com/mperham/dalli/blob/master/lib/dalli/socket.rb

If we did the same thing here, we could refuse to enable the feature unless `BindingOfCaller` is defined.

**Comment by eugeneius** [comment]
Since this feature is optional and disabled by default, we shouldn't monkey patch `Kernel` unless it's going to be used. Could we put the monkey patch in a separate file, and only require it when `capture_locals` is set to `true`?

File: sentry-raven.gemspec (hunk 3 above)
Comment (nateberkopec) [inline sentry-raven.gemspec]: We should re-use [the `integrations` infrastructure](https://github.com/getsentry/raven-ruby/blob/master/lib/raven/base.rb#L203) already in the project, which deals better with Gemfile ordering issues.

**Comment by nateberkopec** [comment]
I think `Raven.inject` is called before the configuration is loaded, but there may be another way to go about it.

**Comment by if1live** [comment]
I modify raven-ruby to support collecting local variables, and I use it 3 months.
I found that collected local variables are broken when implicit exception raised (ex: 1/0)
I don't know how to fix it. So, I close this is now. If I found a way to fix bug, I will rewrite pull request.

**Comment by Shelvak** [comment]
Hi there, I know that was abandoned, But maybe after almost 2 years you have something similar.
Is there anything that could track the locals in the trace?  I use the python client, and all the locals context is great when you need to debug. Thanks in advance.

**Comment by nateberkopec** [comment]
I'm investigating using FFI to provide this functionality for MRI 2.0+ and maybe JRuby 9k, but no timeline. This feature is only really useful in production, but there are no real production-ready APIs for it, so it's pretty complicated

**Comment by Shelvak** [comment]
Ok, thanks for the info and the answer. 

Cheers ✋

**Pull request closed without merging**